### PR TITLE
[dotnet-watch] Make pipe reads and writes fully async

### PR DESCRIPTION
**Pull request by tmat**
Replaces usage of BinaryReader and BinaryWriter, which do not support async reading/writing.
Instead we access the pipe stream directly. The data encoding is the same as BinaryReader/BinaryWriter, which is validated by tests.

All pipe operations are now async and cancellable.

See https://github.com/dotnet/runtime/issues/17229

### CODE REVIEW

**Comment by tmat** [comment]
@phil-allen-msft ptal

File: src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs [hunk 1]
```diff
@@ -59,7 +59,7 @@ public static void Initialize()
                 agent.Reporter.Report("Writing capabilities: " + agent.Capabilities, AgentMessageSeverity.Verbose);
```
Comment (phil-allen-msft) [inline src/BuiltInTools/DotNetDeltaApplier/StartupHook.cs:59]: This is getting to be a pretty long delegate.  Would it make sense to extract it into a method that takes (it appears) just a CancellationToken?

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs [hunk 2]
```diff
@@ -19,58 +29,32 @@ internal readonly struct UpdatePayload(IReadOnlyList<UpdateDelta> deltas, Respon
...
-            for (var i = 0; i < Deltas.Count; i++)
+            foreach (var delta in Deltas)
             {
-                var delta = Deltas[i];
-                binaryWriter.Write(delta.ModuleId.ToString());
-                await WriteBytesAsync(binaryWriter, delta.MetadataDelta, cancellationToken);
-                await WriteBytesAsync(binaryWriter, delta.ILDelta, cancellationToken);
-                await WriteBytesAsync(binaryWriter, delta.PdbDelta, cancellationToken);
-                WriteIntArray(binaryWriter, delta.UpdatedTypes);
+                await stream.WriteAsync(delta.ModuleId, cancellationToken);
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:37]: Are there places in here where the state of the CancallationToken should be checked?

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs [hunk 3]
```diff
@@ -79,117 +63,369 @@ public static void WriteLog(Stream stream, IReadOnlyCollection<(string message,
...
 
             var deltas = new UpdateDelta[count];
             for (var i = 0; i < count; i++)
             {
-                var moduleId = Guid.Parse(binaryReader.ReadString());
-                var metadataDelta = await ReadBytesAsync(binaryReader, cancellationToken);
-                var ilDelta = await ReadBytesAsync(binaryReader, cancellationToken);
-                var pdbDelta = await ReadBytesAsync(binaryReader, cancellationToken);
-                var updatedTypes = ReadIntArray(binaryReader);
+                var moduleId = await stream.ReadGuidAsync(cancellationToken);
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:77]: Is this an area where checking the CancellationToken is needed?

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs [hunk 4]
```diff
@@ -79,117 +63,369 @@ public static void WriteLog(Stream stream, IReadOnlyCollection<(string message,
...
+        private const byte Version = 1;
+
+        public string AssemblyName { get; } = assemblyName;
+        public bool IsApplicationProject { get; } = isApplicationProject;
+        public string RelativePath { get; } = relativePath;
+        public byte[] Contents { get; } = contents;
+
+        public async ValueTask WriteAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            await stream.WriteAsync(Version, cancellationToken);
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:152]: Are there comments somewhere that describe how the bytes are supposed to be laid out?

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs [hunk 5]
```diff
@@ -79,117 +63,369 @@ public static void WriteLog(Stream stream, IReadOnlyCollection<(string message,
...
+                b = await stream.ReadByteAsync(cancellationToken);
+                result |= (b & 0x7Fu) << shift;
+
+                if (b <= 0x7Fu)
+                {
+                    return (int)result;
+                }
+            }
+
+            // Read the 5th byte. Since we already read 28 bits,
```
Comment (phil-allen-msft) [inline src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:400]: Could this run across different versions, where the code herein has a different version between the "client" and the "server"?  If so, would there need to be tests with the serialized state, where that serialized state is checked in and effectively unchanging?

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs (hunk 2 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:37]: You mean explicitly calling cancellationToken.ThrowIfCancellationRequested? There is no expensive work done in these helpers they are just calling stream.WriteAsync

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs (hunk 3 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:77]: Not sure why would it be needed. We are calling to Stream.ReadAsync, which will use the cancellation token while waiting for the data.

File: src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs (hunk 4 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs:152]: It can be anything as long as we have the same code on both sides.

**Review by phil-allen-msft — APPROVED** [review]

**Pull request merged**